### ISSUE TITLE
Standalone version (for Linux) with NSM support

### ISSUE THREAD

**Issue by grammoboy2** [surge-synthesizer/surge#3515]
Thx, interesting synth. Worth to release as standalone application probably. 

So a feature request:

Standalone application (for Linux) with NSM support.

Plugins can be nice, but I actually like it when I have my synth open in a dedicated window / desktop space, instead of launching the plugin in track in a DAW and find myself just picking a preset, because the plugin overlaps my track in the DAW. 

Of course this is somewhat possible with Carla, but still you need a plugin host to rely on. 

Synths like Zynaddsubfx do have Non/New-Session-Manager (NSM) support and they work nice as standalone application that way. 

https://non.tuxfamily.org/wiki/Non%20Session%20Manager
http://non.tuxfamily.org/nsm/API.html
https://github.com/linuxaudio/new-session-manager

**Comment by mkruselj** [comment]
Pull requests for such a thing are welcome 🙂

**Comment by baconpaul** [comment]
I am surprised someone hasn't written an 'here's an nsm host for a vst3'.
But I'd never heard of NSM before so what do I know!
No plans for a standalone any time soon but as @mkruselj happy for someone who is inspired to do so.

**Comment by mkruselj** [comment]
Oh so NSM is done through OSC? Surge would have to [support OSC](https://github.com/surge-synthesizer/surge/issues/2355) first then. Indeed not happening any time soon.

**Comment by grammoboy2** [comment]
NSM implementation is relatively easy. You don't have to have full OSC support.
:switch: functionality, is maybe most of the work, but it also works without.
See API.

**Comment by baconpaul** [comment]
Cool! I’m glad it is easy! If you want to design a wrapper for surge that uses it happy to provide pointers on cmake, Submodules etc. probably a wrapper symmetric to the au vst lv2 headless and python bindings is my guess

**Comment by grammoboy2** [comment]
Ha, this is what people tell me. The hardest part is reading the API and follow the rules it seems... There I can help, but I'm not a programmer.

**Comment by baconpaul** [comment]
Ahh gotcha

Well look this probably isn’t going to make our list in 2021 for the current devs (who are all daw users and already have a big set of plans for the plugin) but if you can recruit a dev inspired to do it or if someone reading this you otherwise wants to i am happy to help them navigate the code and the possibilities!

**Comment by gnac** [comment]
There is a standalone executable that could be put into a package. It wouldn't necessarily use NSM, but it does work with ALSA and Jack as is. Would that be an interim step?

**Comment by grammoboy2** [comment]
Well, that sounds like an important step, great. JACK is the most interesting to use with for low latency audio, for me at least. Step by step, but a stable standalone version is a requirement before even thinking about NSM support.

**Comment by baconpaul** [comment]
We are now getting a stand-alone through juce yes and our alpha installer includes the binary of it but is still very alpha 

I have no idea of or how juce would support nsm

**Comment by grammoboy2** [comment]
How to implement NSM support in a client is described in the following links. It's not implemented magically, by using a toolkit like Juce. 

The idea is that a NSM manager manages all the standalone applications in a session. For that to work, a client needs to have NSM support and comply to the rules of NSM when the applications is launched within a NSM session (when the NSM client has announced itself as NSM client to the running NSM manager via simple OSC). The rules makes sure all clients behave the same way and all store their data in a session folder. A little client 'jackpatch' is saving the JACK connections of a particular NSM session (this is not the concern of the client). 

The nice thing about JACK it it's flexibility, but to startup and save all individual clients and connections can be a burden, so that's why NSM was invented. 

https://non.tuxfamily.org/wiki/Non%20Session%20Manager

"The API comprises a simple Open Sound Control (OSC) based protocol, along with some behavioral guidelines, which can easily be implemented by various applications. "

http://non.tuxfamily.org/nsm/API.html

The hard thing for devs seems to be to implement the strict rules, but that's mostly a matter of reading the API carefully I would think. There is also a :switch: feature described, which is probably the most difficult to implement, but that is just nice to have, it's not strictly needed to function. That feature could be added later.

**Comment by jpcima** [comment]
There is my custom [standalone with NSM], this code is free to reuse.
I imagine it would be straightforward to port another synth on it.

Although Juce has its own jack audio client, it is not very well made (imo) and doesn't interact with the system in a way that typical Jack apps do.

[standalone with NSM]: https://github.com/jpcima/ADLplug/blob/master/sources/jack_main.cc

**Comment by baconpaul** [comment]
Sounds like maybe doing the same thing the lv2 folks did to make a juce fork that makes juce apps work with nsm would be a reasonable investment for you all? Surge xt uses production mainline juce but can build against alternates which is how we get the lv2

**Comment by grammoboy2** [comment]
NSM support for all JUCE apps would be great, I would say as NSM user. I prefer a synth as dedicated standalone application above a plugin. Next to Surge, Vital(ium) would be another nice candidate as Juce standalone with NSM support.

A second possibility might be to add NSM support to Jalv (which then should run with a GUI at start), which is a simple LV2 host that says it can run LV2 plugins as JACK standalone applications. But I would still prefer a standalone application with NSM support, without a extra plugin layer. 
http://drobilla.net/software/jalv

**Comment by mkruselj** [comment]
You would need to pitch that to JUCE team, but the reality is it's quite unlikely they'd add it.

**Comment by baconpaul** [comment]
Or maintain a fork. Or create a juce module which implements it. I don’t know the spec but maybe a module could get you there. And the lv2 crew maintains a fork for lv2 juce.

**Comment by mkruselj** [comment]
If JUCE ends up supporting this, so will Surge XT. If they don't, we won't, unless somebody sends us a pull request.

**Comment by grammoboy2** [comment]
> If JUCE ends up supporting this, so will Surge XT. If they don't, we won't, unless somebody sends us a pull request (at which point reopen this issue please).

The problem seems that neither JACK support, nor the OSC support in @juce-framework is optimal. At least, I see problems with JACK and BespokeSynth. Also, the OSC library of JUCE isn't able to reply to sender as far as I can see. @baconpaul has solved this for his application to implement JACK and OSC (not using JUCE osc, but the liblo library) himself.

In general I can imagine that removing the need for a LV2 host from the pipeline, reduces the likelihood of running into issues when using Surge. Now a user can have potential issues in Surge, but also in the LV2 host. Also one would think that less system resources would be needed, without the need for a LV2 host.

**Comment by baconpaul** [comment]
(1.3 is the currently unreleased head of code - the nightly has osc but we have some message types to add before completion)

**Comment by grammoboy2** [comment]
> In 1.3 we added osc support including bidirectional conversation. If that helps.

That's interesting. So you're not using the JUCE OSC library then?

The main question then is how Surge XT is performing with JACK. I can't get it to run, CPU goes up to 70%, but the Surge XT GUI doesn't appear, while it does when I stop JACK.

If I understand it correctly, Surge is using JUCE for it's standalone capability and the JACK implementation in JUCE is far from optimal it seems. Both Vital and Surge and also BespokeSynth are high quality synths, which could be good candidates for a dedicated standalone version with NSM support, but as long those don't perform well with JACK (via JUCE), it will be a waste effort. It looks like attempts to improve the JUCE / JACK combination are somewhat neglected by the JUCE team. 

Then a other option would be to implement JACK outside of JUCE like @jpcima has done.

It's not that I see this happening any time soon, but it got me thinking anyway.

**Comment by baconpaul** [comment]
You just open and handle two ports basically and you can do it, a bit of irritation but doable,

it’s pretty easy to write a juce standalone using rtaudio except for the gui. We did something very similar in the cli surge. If you are interested in doing that I can give you pointers. Starting a gui thread without juce is painful but you could also do audio with rtaudio rtmidi and gui and startup main with juce easily. Just look in src/surge-xt/cli for what we’ve done with true headless. 

I’m not gonna do this but if someone did I would happily merge it as an alternate standalone

**Comment by mkruselj** [comment]
We're using JUCE not just for standalone, but also for the GUI and all the plugin abstractions.

**Comment by baconpaul** [comment]
Yeah but those could be easily wrapped in another audio and midi source. That’s basically what the cli does (just it uses juce as the audio and midi source). 

So what I was proposing is use juce main for the window and stuff but use rtaudio for the connectivity instead of juce::audiomanager - the cli shows how to write the callbacks

**Comment by grammoboy2** [comment]
Thanks. The best thing would still that @juce-framework would fix their JACK issues. Then it would applicable to Vital, BespokeSynth etc as well. But if those devs are not really open to it, the way to do it as suggested here, might be a better way indeed. I don't see myself capable of doing it atm, but who knows in the future.

**Comment by baconpaul** [comment]
Has anyone given a clear explanation of what those issues are on the juce forum? If there’s a forum post which lays it out clearly and politely I’m happy to give it a ping. 

The limit of my jack knowledge is: 1) I can turn it on with cmake, 2) I can start it when I’m in Linux and 3) I conceptually understand it but I have core audio on my Mac so don’t need to actually understand it.

**Comment by grammoboy2** [comment]
Thanks for your interest. When I find a moment, I'll do some 'research' regarding JUCE and JACK.

**Comment by baconpaul** [comment]
Sharing it on the juce forum is the most effective but when you do please share the link and I’ll happily share thst it relates to your use of our synth
